### ISSUE TITLE
[DOC] In-Depth README

### ISSUE THREAD

**Issue by rosadev123** [rosadev123/rosa_dashboard#1]
# README for ROSA Dashboard

## 1. Introduction
- **Welcome**: This README provides a comprehensive guide for developers tasked with maintaining, updating, or expanding the Rosa dashboard project. It is designed to provide all necessary details about the project’s setup, structure, and workflow, ensuring a smooth handoff or continuation of development.
- **Audience**: This document is intended for developers with a strong understanding of:
  - **R**: Knowledge of its syntax, package management, and data handling.
  - **Quarto**: Understanding its role in creating dynamic documents and dashboards.
  - **Targets Package**: Familiarity with Make-like pipeline tools that optimize workflows by managing dependencies and automating updates.
    - **Targets Documentation**: The official [targets package guide](https://books.ropensci.org/targets/) is an excellent resource for understanding its structure and usage.
    - **Make-like Workflow**: Targets tracks dependencies, automates redundant processes, and integrates parallel computing, making it ideal for complex analysis pipelines.
- **Scope**: This README covers project setup, workflow, customization, version control, and access to supplemental materials.

### Important Links

https://github.com/rosadev123/rosa_dashboard
Google Credentials - see section  

---

## 2. Project Overview
- **Purpose**: The Rosa dashboard is an interactive tool that allows stakeholders to explore and filter research needs and projects. The goal is to streamline data access and improve the decision-making process by visualizing data in an intuitive format.
- **Features**:
  - Interactive filters for drilling down into data subsets.
  - Visualizations and summaries to present research projects and needs effectively.
  - A self-contained HTML document that functions independently in any modern browser.
- **Technology**:
  - **R**: The primary programming language for data processing and dashboard generation.
  - **Quarto**: Used to structure and style the HTML dashboard.
  - **Targets Package**: Manages dependencies, automates processing, and tracks the workflow.

---

## 3. Environment Setup

Setting up the project environment is a critical step to ensure that the Rosa dashboard can be maintained, updated, and reproduced reliably. 

This project uses the `renv` package to manage dependencies, allowing you to replicate the exact package versions used during development. 

Additionally, two custom R packages, `gauntlet` and `gauntletReactable`, are required but are not available on CRAN. These custom packages can be installed from the provided source files or from the G Drive.

Follow the steps below to configure your environment and install all necessary dependencies.

---

### **Required Software**
- **R**:
  - Version 4.3.0 or higher is recommended for compatibility with the dependencies and features used in this project.
- **RStudio**:
  - RStudio is the recommended IDE for working with this project due to its seamless integration with R and Quarto.
- **Quarto**:
  - Version 1.3.0 or higher is required for rendering the dashboard from the `.qmd` file.

---

### **Dependencies**

This project uses the `renv` package to manage CRAN packages and their versions. Additionally, it requires two custom packages, `gauntlet` and `gauntletReactable`, which are not available on CRAN.

 Official documentation: [renv Package](https://rstudio.github.io/renv/articles/renv.html)

#### **What is `renv`?**

- **Overview**:
  - `renv` is a package management tool for R that ensures reproducibility by isolating project-specific libraries. It captures the exact package versions used in your project and allows others to restore the same environment effortlessly.

- **Key Features**:
  - Creates a project-specific library (`renv/library/`) separate from your global R library.
  - Saves package dependencies in a `renv.lock` file, enabling others to replicate the environment.
  - Allows you to snapshot (`renv::snapshot()`) and restore (`renv::restore()`) the environment to ensure consistency.

---

#### **Installing CRAN Dependencies**

1. **Automatic Installation**:
   - Use the `renv::restore()` command to install all CRAN dependencies as specified in the `renv.lock` file.
   - Example:
     ```R
     renv::restore()
     ```

2. **List of Core CRAN Packages**:
   - All CRAN packages will be loaded using the above step.

3. **If Installation Fails**:
   - Ensure that CRAN repositories are accessible from your system.
   - Check for missing or outdated packages using:
     ```R
     renv::status()
     ```
- The `renv.lock` file lists all packages and their versions, this can be used for manual install if need be

---

#### **Installing Custom Packages**

This project relies on two custom packages, `gauntlet` and `gauntletReactable`, which are not available on CRAN. These packages are essential for generating and interacting with the dashboard.

1. **Installing from G Drive**:
   - Both packages are provided as zip files in the **special_packages** folder of the G Drive (refer to Section 10).
   - These packages were archived as of **2024 December 3**. You can download the zip files and install them locally:
     ```R
     install.packages("path/to/gauntlet.zip", repos = NULL, type = "source")
     install.packages("path/to/gauntletReactable.zip", repos = NULL, type = "source")
     ```

2. **Installing from GitHub (Optional)**:
   - Placeholder links for direct GitHub installation (replace with actual URLs):
     - `gauntlet`: [https://github.com/michaelgaunt404/gauntlet](https://github.com/michaelgaunt404/gauntlet)
     - `gauntletReactable`:  [https://github.com/michaelgaunt404/gauntletReactable](https://github.com/michaelgaunt404/gauntletReactable)
   - To install from GitHub using `remotes`:
     ```R
     devtools::install_github("michaelgaunt404/gauntlet")
     devtools::install_github("michaelgaunt404/gauntletReactable")
     ```

3. **Validating Installation**:
   - After installation, load the packages to confirm they are working correctly:
     ```R
     library(gauntlet)
     library(gauntletReactable)
     ```
---

## 4. File Structure
The file structure of this project is organized to support efficient navigation and clear separation of responsibilities. Each file and folder serves a specific purpose, enabling developers to quickly locate and update relevant components without confusion. This section provides an overview of the key files and folders, along with guidance on their intended use. Understanding this structure is crucial for maintaining the pipeline and ensuring the dashboard functions as expected.

### Files
The project’s core files are critical for configuring the environment, managing the pipeline, and generating the dashboard. While some files, such as the _quarto.yml and _targets.R, require occasional updates or modifications, others, like README.md or rosa_dashboard.Rproj, are static and provide documentation or project-specific settings. This subsection outlines each file's purpose to help developers understand their roles and dependencies.

Below is a representation of the repo and its structure:
```
Project Root Directory
├── _quarto.yml          # Quarto configuration file
├── _targets.R           # Targets pipeline configuration
├── README.md            # Project documentation
├── rosa_dashboard.Rproj # RStudio project file
├── run.R                # Script for local workflow testing
├── run.sh               # Shell script for automation
├── _targets/            # Targets output directory (auto-generated)
│   ├── objects/         # Contains pipeline outputs
│   └── logs/            # Logs of pipeline runs
├── analysis/            # Quarto document and dashboard outputs
│   ├── dashboard.qmd    # Main Quarto file defining the dashboard
│   └── dashboard.html   # Rendered HTML output
├── code/                # Scratch scripts for exploratory work
│   └── example_script.R # Example exploratory script
├── data/                # Input data files
│   └── rosa_database.xlsx # Latest database provided by Rosa
├── docs/                # Supplemental documentation
│   └── process_notes.md # Notes or additional guides
├── R/                   # Custom R functions
│   ├── process_data.R   # Function for processing data
│   └── generate_plot.R  # Function for creating visualizations
```
1. **_quarto.yml**: Configures the dashboard’s layout and structure, such as theme, navigation, and file paths.
2. **_targets.R**: The core pipeline file that defines data processing and output generation. This file contains all target definitions and their dependencies.
3. **README.md**: The main documentation file explaining project setup and maintenance.
4. **rosa_dashboard.Rproj**: The RStudio project file for managing the R environment.
5. **run.R**: A script for testing and executing parts of the pipeline locally.
6. **run.sh**: A shell script for automating repetitive tasks like rendering the dashboard.

### Folders
The folder structure is designed to organize resources, outputs, and supplementary materials in a logical and maintainable way. Key folders include `./data` for the input data, `./analysis` for the Quarto dashboard and its output, and `./R` for reusable custom functions. This subsection explains the role of each folder and highlights best practices for interacting with them during development and maintenance.

1. **./_targets**:
   - Stores intermediate and final outputs from the `targets` workflow.
   - Avoid manual interaction unless troubleshooting.
2. **./analysis**:
   - Contains the Quarto `.qmd` file defining the dashboard.
   - Stores the rendered HTML dashboard output.
3. **./code**:
   - Houses exploratory or temporary scripts.
   - Not intended for reusable functions or production code.
4. **./data**:
   - Stores the most recent Excel workbook from Rosa.
   - Archive older versions in the G Drive before replacing.
5. **./docs**:
   - Supplemental documents unrelated to dashboard generation.
   - May include process overviews, notes, and troubleshooting guides.
6. **./R**:
   - Contains all reusable R functions.
   - Functions here are automatically sourced during pipeline execution via `tar_source()`.

---

## 5. Data Workflow

The data workflow for the Rosa dashboard project is designed to be as automated and streamlined as possible, ensuring consistency and minimizing the potential for errors during data updates. The workflow integrates the `targets` package for pipeline management, allowing the ingestion, transformation, and visualization of data to be handled efficiently. 

This section outlines the source of the data, the steps required to update or add new data, and the automated processes that transform the raw input into the final dashboard outputs. Understanding this workflow is essential for maintaining and extending the project while adhering to the structure and conventions established by the original developer.

Key topics include:
- **Data Source**: Details about the format and structure of the data provided by Rosa.
- **Automated Processes**: How the pipeline manages and processes data.
- **Developer Responsibilities**: What steps need to be taken manually and what the system automates.

By following the guidelines in this section, developers can ensure that the dashboard remains accurate, up-to-date, and aligned with the client's expectations.

- **Data Source**:
  - Rosa provides a regularly updated Excel workbook containing multiple tabs:
    1. **Research Projects**: Metadata on current projects.
    2. **Research Needs**: A comprehensive list of identified gaps.
    3. **Summarized Research Needs**: Aggregated mappings connecting projects and needs.
    4. **References**: Definitions, acronyms, and other supplemental information.
- **Updating the Data**:
  1. Archive the current workbook to the designated folder in the G Drive.
  2. Copy the new workbook into the `data` folder within the project repository.
  3. Update `_targets.R` by modifying the file path for the `data_rosa_dbase_file` target.
- **Workflow Automation**:
  - The `targets` package processes the workbook, extracting and transforming data based on predefined pipelines.
  - Data ingestion, transformation, and visualization are fully automated, reducing the potential for human error.
- **Preprocessing**:
  - Minimal preprocessing is needed manually; the pipeline automates most tasks.
  - Any additional preprocessing must be added to the relevant target functions in `_targets.R`.

---

## 6. Quarto Document Details
- **File Overview**:
  - The `.qmd` file in `./analysis` defines the dashboard structure and layout.
  - It uses Quarto’s enhanced markdown syntax to integrate R outputs directly into the HTML dashboard.
- **Customization**:
  - Filters and visualizations can be modified by editing the `.qmd` file.
  - CSS and JavaScript can be referenced in `_quarto.yml` for advanced styling.
- **Resources**:
  - Learn more about Quarto from its [official documentation](https://quarto.org/).

---

## 7. Updating the Dashboard

Keeping the dashboard up to date is a critical part of this project. 

The update process is designed to be as automated as possible using the `targets` pipeline. This section outlines the complete steps for updating the dashboard when new data is provided or when changes to the dashboard structure are required. 

By following these steps, developers can ensure the workflow runs smoothly, dependencies are managed correctly, and the dashboard reflects the most recent data and design specifications.

Below is a description of a typical dashboard update cycle.

### Steps to Update the Dashboard
1. **Receive Updated Data**:
   - Rosa will provide an updated Excel workbook, typically containing new research projects, updated research needs, or modified metadata.
   - Confirm the validity and completeness of the data provided by Rosa before proceeding.

2. **Archive the Current Workbook**:
   - Navigate to the designated Google Drive folder used for archiving past versions of the workbook.
   - Save the current workbook to this folder with a clear version identifier (e.g., `rosa_database_2023-12-01.xlsx`).

3. **Replace the Workbook in the Project Repository**:
   - Copy the new workbook into the `data` folder located in the project directory.
   - Use a consistent naming convention for the new file (e.g., `rosa_database_2024-01-15.xlsx`).

4. **Update the Pipeline Configuration**:
   - Open the `_targets.R` file in the root directory.
   - Update the `data_rosa_dbase_file` target to point to the new file path:
     ```R
     tar_target(
    data_rosa_dbase_file
    ,here::here(
      "data"
      ,"FishFORWRD v2.2.0_8.14.24_mg.xlsx"    #change this pointer
    ), format = "file")
     ```
   - Save the `_targets.R` file to finalize the update.

5. **Visualize Dependencies**:
   - Use the `tar_visnetwork()` function to review the targets pipeline. This visualization will show which targets are affected by the updated data and need to be rebuilt.
   - Below is an example of what the target network should look like when the input database is changed.
     - All functions are loaded (green triangles)
     - All data objects depend on `data_rosa_dbase_file` are out-of-date (blue circles)

![Image](https://github.com/user-attachments/assets/e5e048fa-c703-4d4d-9c5d-4d246a580aa4)
 
6. **Rebuild the Pipeline**:
   - Run the `tar_make()` function to execute the pipeline. This process:
     - Sources all functions from the `R/` folder.
     - Ingests and processes the new data.
     - Regenerates intermediate targets, visualizations, and the final dashboard.
   - Monitor the process for errors or failed targets. If any errors occur, troubleshoot the specific function or target as needed.
   - `tar_make()` is a noisy function and will print in real-time which targets it is currently working and their status, see example below:
   
![Image](https://github.com/user-attachments/assets/dfa4593a-d497-495e-9481-1b9969b8333f)

7. **Validate the Dashboard Output**:
   - Navigate to the `analysis` folder and open the newly generated `dashboard.html` file.
   - Review the dashboard to ensure:
     - Data is accurate and up to date.
     - Visualizations render correctly without errors.
     - Filters and interactive elements function as intended.

8. **Deliver the Updated Dashboard**:
   - Once validated, share the updated HTML file with Rosa’s web team for integration into their website.
   - Provide any necessary notes or change summaries to help the Rosa team understand updates or improvements.

### Additional Notes
- **Troubleshooting**:
  - If the pipeline fails, use `tar_visnetwork()` to identify the specific targets causing issues.
  - Debug the relevant functions by loading their dependent targets using `tar_load()` and testing them step-by-step.
  - Below is an example of a targets network with a failed target build

![Image](https://github.com/user-attachments/assets/2e0f86c5-80c8-49d7-9a51-25d89b50598c)

- **Best Practices**:
  - Always archive the previous version of the data before replacing it.
  - Document any significant changes in the pipeline or dashboard structure in the project’s `README.md` or `docs` folder.
- **Automation**:
  - The provided `run.R` or `run.sh` scripts can automate parts of this process, such as running the pipeline and rendering the dashboard.

---

## 8. Customization and Updates
- **Feature Development**:
  - Use Git branches to isolate and test new features.
  - Save all reusable R functions in the `R` folder to ensure automatic sourcing.
- **Best Practices**:
  - Maintain clear and consistent naming conventions for files and functions.
  - Document any significant changes in the `README` or supplemental notes.
  
---

## 9. Version Control

Version control is critical for maintaining the integrity and continuity of this project. The use of Git ensures that updates—whether simple data refreshes or feature enhancements—are tracked and managed effectively. Given the nature of the project, version control becomes particularly important for several reasons:

1. **Long Gaps Between Updates**:
   - Updates to the dashboard may occur sporadically, sometimes with months or even years between them. Using Git allows developers to revisit the project with a clear history of changes, minimizing time spent relearning or understanding past modifications.

2. **Safeguarding the Existing Codebase**:
   - Git provides a safety net for the existing codebase, ensuring that all updates are incremental and reversible. By using branches and version tags, developers can experiment with changes without risking the stability of the production version.

3. **Collaboration Across Multiple Developers**:
   - In cases where multiple developers are involved, or when handoffs occur, Git serves as a central repository of truth. Even if direct communication between team members is limited, the commit history, branches, and pull requests provide a clear record of contributions and rationale for changes.

### Tools for Version Control

To effectively use version control, you will need to install Git and optionally a Git client for easier visualization and management. Recommended tools:

- **Git**:
  - Website: [Git](https://git-scm.com/)
  - Description: A free, open-source distributed version control system for tracking changes in code and collaborating on projects.
- **GitKraken**:
  - Website: [GitKraken](https://www.gitkraken.com/)
  - Description: A popular Git client with an intuitive graphical interface for managing repositories, branches, and pull requests.

Both tools integrate seamlessly with GitHub, the hosting platform for this project’s repository.

### Git Workflow for the Project

1. **Clone the Repository**:
   - Begin by cloning the project repository to your local machine:
     ```bash
     git clone https://github.com/rosadev123/rosa_dashboard
     ```
   - Ensure you have the correct access credentials (refer to Section 10 for account details).

2. **Create a New Feature Branch**:
   - Always create a new branch for updates or feature development to avoid directly modifying the `main` branch:
     ```bash
     git checkout -b feature/update-data-2024
     ```
   - Use descriptive branch names that reflect the purpose of the update (e.g., `feature/new-filter`, `bugfix/fix-plot-error`).

3. **Commit Changes Frequently**:
   - Commit changes regularly with meaningful commit messages. Each commit should capture a logical, self-contained update:
     ```bash
     git add .
     git commit -m "Updated data pointer to new Rosa database"
     ```
   - Avoid large, single commits that combine multiple unrelated updates.

4. **Push Changes to Remote Repository**:
   - Push your branch to the remote repository to ensure your work is backed up and accessible to collaborators:
     ```bash
     git push origin feature/update-data-2024
     ```

5. **Submit a Pull Request (PR)**:
   - When the updates are complete and tested, submit a PR to merge your branch into the `main` branch:
     - Clearly describe the changes in the PR title and description.
     - Assign reviewers, if applicable,

---

## 10. Non-R-Based Supplemental Materials

The Rosa dashboard project is supported by a Google account that provides access to critical supplemental materials, including archived data, documentation, and custom R packages. Additionally, the associated GitHub repository is essential for version control and collaborative development. This section details the structure and usage of these resources.

###  Google Account

The shared Google account serves as the central hub for accessing archived data, documentation, and project-specific resources. This account ensures continuity and availability of resources for all developers working on the project.

- **Account Details**:
  - Email: `rosa.developer.acc@gmail.com`
  - Password: [[REDACTED]]
- **Best Practices**:
  - Ensure that sensitive credentials are not shared openly or included in public repositories.
  - Update the password periodically for security and document the new credentials in the appropriate internal documentation.

### Google Drive Structure

The Google Drive associated with the shared account is organized into three primary folders, each serving a specific purpose. This structure ensures efficient management and accessibility of resources.

#### Data Folder
- **Purpose**: Archives all versions of the Rosa database.
- **Current Content**: All data available as of December 3, 2024, has been saved here by the original developer, Michael Gaunt.
- **Usage Guidelines**:
  - Save new versions of the Rosa database in this folder.
  - Use consistent naming conventions to include version numbers and dates (e.g., `rosa_database_v2.2.0_2024-12-03.xlsx`).
  
#### Docs Folder
- **Purpose**: General-purpose storage for project-related documentation.
- **Subfolders**:
  - **archive_2024-12-03**: Contains documents available on Michael Gaunt’s local drive at the time of this document’s creation.
    - Includes a PowerPoint presentation explaining the `targets` package and its workflow management capabilities.
- **Usage Guidelines**:
  - Add relevant documentation to this folder to maintain a comprehensive archive.
  - Avoid duplicating materials unnecessarily between the G Drive `docs` folder and the repository `docs` folder.

#### Special_Packages Folder
- **Purpose**: Contains zip files of custom R packages used in the Rosa dashboard development.
- **Usage Guidelines**:
  - Install packages from these source files if they are not available in your local environment.
  - Archive updates to the custom packages here for future reference.

#### ASCII Representation of the Google Drive Structure

```
Google Drive Root
├── data/                     # Archives all Rosa database files
│   ├── rosa_database_v2.1.0_2024-11-15.xlsx    --just and example
│   ├── rosa_database_v2.2.0_2024-12-03.xlsx    --just and example
├── docs/                     # Supplementary and archived documentation
│   ├── archive_2024-12-03/   # Archive of original developer's local documents
│   │   ├── targets_workflow_guide.pptx
│   │   └── additional_notes.docx
├── special_packages/         # Non-CRAN R packages required for the project
│   ├── gauntlet.zip
│   ├── gauntletReactable.zip
```

### GitHub Repository

The GitHub repository associated with the shared account is used to manage the project’s codebase and facilitate collaborative development. Proper use of this repository ensures that all changes are tracked and the codebase remains stable.
`sdiud!@ssl_github`
- **Repository Details**:
  - Email: `rosa.developer.acc@gmail.com`
  - Password: [[REDACTED]]
  - Repository URL: `https://github.com/rosa-dev/rosa-dashboard.git`

#### Key Features of the Repository
1. **Branch Management**:
   - The `main` branch is reserved for the stable, production-ready version of the project.
   - Use feature branches for new development and bug fixes to isolate changes and facilitate collaboration.

2. **Pull Requests**:
   - All updates must be reviewed and approved via pull requests before being merged into the `main` branch.
   - Use descriptive titles and summaries in pull requests to document the purpose and scope of changes.

3. **Versioning**:
   - Tag major updates with version numbers for easier tracking of project milestones.
   - Example of tagging a release:
     ```bash
     git tag -a v1.0 -m "Initial dashboard release"
     git push origin v1.0
     ```

#### Best Practices for Using the Repository
- Commit changes frequently with clear, descriptive messages.
- Regularly pull changes from the remote repository to stay updated with the latest developments.
- Document significant changes in the repository’s `README.md` or create a `CHANGELOG.md` file.

### **GitHub Repository: Project Board**

In addition to hosting the codebase, the GitHub repository for the Rosa dashboard includes a **Project Board** designed to track features, bugs, and other tasks. The Project Board is an essential tool for maintaining continuity across developers and ensuring that work remains organized and prioritized.

#### **Purpose of the Project Board**
- **Feature Tracking**:
  - Lists all new features under development, along with descriptions and progress.
- **Bug Reporting**:
  - Tracks identified bugs, their severity, and their current resolution status.
- **Task Prioritization**:
  - Organizes tasks by priority to ensure critical issues and features are addressed first.
- **Developer Handoff**:
  - Ensures that information about ongoing and completed tasks persists across developers, reducing onboarding time and preventing duplicated work.

#### **Structure of the Project Board**
The Project Board uses a **Kanban-style layout** with the following tabs:
1. **To Do**:
   - Contains all unstarted tasks or ideas for features and bug fixes.
2. **In Progress**:
   - Tracks tasks actively being worked on by developers.
3. **Done**:
   - Lists all completed tasks, providing a history for reference.
4. **Holding Pattern**:
   - Contains tasks or issues that are not actively being worked on but may require attention in the future (e.g., blocked by dependencies or awaiting further input).
5. **Commemoration**:
   - A space to celebrate significant milestones, completed features, or achievements related to the project.

#### **How to Use the Project Board**
1. **View Tasks**:
   - Navigate to the Project Board from the GitHub repository's main page by clicking on the "Projects" tab.
2. **Add a New Task**:
   - Create a new card for a feature or bug by clicking the "+" button in the appropriate column.
   - Include a clear title and a description of the task or issue.
3. **Move Cards**:
   - Update the status of tasks by dragging and dropping cards between columns.
4. **Collaborate**:
   - Use comments on cards to communicate updates, clarify issues, or assign tasks.

#### **Best Practices**
- **Regular Updates**:
  - Keep the board updated with the latest status of tasks to provide a clear picture of progress for all developers.
- **Detailed Descriptions**:
  - Include sufficient detail in cards to ensure the task is understandable to new developers.
- **Use "Holding Pattern" Wisely**:
  - Add tasks to this tab only when there is a clear reason for delay or dependence on external factors.
- **Highlight Wins in "Commemoration"**:
  - Use the commemoration tab to document major milestones or achievements, fostering a sense of accomplishment and history for the project.
- **Cross-Reference Commits**:
  - Link related GitHub issues, pull requests, or commits to the task card for easier navigation and tracking.

By effectively using the Project Board and its tabs, developers can maintain a consistent and transparent workflow, ensuring that the Rosa dashboard project remains organized, prioritized, and on track.
